courses/codepeer/010_codepeer.rst

```diff
@@ -0,0 +1,2666 @@
+   end Incorrect;
+
+| ``medium warning: suspicious constant operation X/15 always evaluates to 0``
+
+--------------------
+Subp Never Returns
+--------------------
+
++ Subprogram will **never** return
+
+  + presumably **infinite loop**
+
++ Typically, **another message** in the body can explain why
+
+  + eg. ``test always false``
+
+..
+   :toolname:`CodePeer` example (4.1.4 - subp never returns)
+
+.. code:: Ada
+   :number-lines: 1
+
+   procedure Infinite_Loop is
+      X : Integer := 33;
+   begin
+      loop
+         X := X + 1;
+      end loop;
+   end Infinite_Loop;
+
+| ``medium warning: subp never returns: infinite_loop``
+
+-------------------
+Subp Always Fails
+-------------------
+
++ A run-time problem could occur on **every** execution
++ Typically, **another message** in the body can explain why
+
+..
+   :toolname:`CodePeer` example (4.1.4 - subp always fails)
+
+.. code:: Ada
+   :number-lines: 1
+
+   procedure P is
+      X : Integer := raise Program_Error;
+   begin
+      null;
+   end P;
+
+| ``high warning: subp always fails: p fails for all possible inputs``
+
+-------------------
+Same Operands
+-------------------
+
++ The two operands of a binary operation are syntactically equivalent
++ The resulting expression will always yield the same value
+
+.. code:: Ada
+   :number-lines: 1
+
+   function Same_Op (X : Natural) return Integer is
+   begin
+      --  Copy/paste error? Always return 1
+      return (X + 1) / (X + 1);
+   end Same_Op;
+
+| ``medium warning: same operands (Infer): operands of '/' are identical``
+
+-------------------
+Same Logic
+-------------------
+
++ The same sub-expression occurs twice in a boolean expression
++ The entire expression can be simplified, or always return the same value
+
+.. code:: Ada
+   :number-lines: 1
+
+   function Same_Logic (A, B : Boolean) return Boolean is
+   begin
+      return A or else B or else A;
+   end Same_Logic;
+
+| ``medium warning: same operands (Infer): 'A' duplicated at line 3``
+
+-------------------
+Test duplication
+-------------------
+
++ The same expression is tested twice in successive :ada:`if ... elsif ... elsif ... `
++ Usually indicates a copy-paste error (CWE 1041)
+
+.. code:: Ada
+   :number-lines: 1
+
+   procedure Same_Test (Str : String) is
+      A : constant String := "toto";
+      B : constant String := "titi";
+   begin
+      if Str = A then
+         Ada.Text_IO.Put_Line("Hello, tata!");
+      elsif Str = B then
+         Ada.Text_IO.Put_Line("Hello, titi!");
+      elsif Str = A then
+         Ada.Text_IO.Put_Line("Hello, toto!");
+      else
+         Ada.Text_IO.Put_Line("Hello, world!");
+      end if;
+   end Same_Test;
+
+| ``medium warning: same test (Infer): test 'Str = A' duplicated at line 9``
+
+-------------------
+Duplicate branches
+-------------------
+
++ Branches are duplicated in a :ada:`if` or a :ada:`case`
++ Should be refactored, or results from incorrect copy-paste (CWE 1041)
```

Comment:
This is our first actual reference to a CWE. We reference the acronym in "Typical Users and Use Cases", but don't explain it at all

Reply:
Removed CWE numbers